src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala

```diff
@@ -49,38 +56,81 @@ class HttpGoogleCloudStorageDAO(
+          _ match {
+            case Success(g) => g
+            case Failure(t) => throw t
+          }
+        }
+      }
+    }
+
+    def insertOwnerMember: (Seq[Group]) => Future[Member] = { _ =>
+      retry(when500) {
+        () => Future {
+          val ownersGroupId = toGroupId(bucketName, WorkspaceAccessLevel.Owner)
           val owner = new Member().setEmail(userInfo.userEmail).setRole(groupMemberRole)
-          val ownerInserter = directory.members.insert(ownersGroupId,owner)
-          retry(ownerInserter.execute,when500)
-
-          val readersEntity = makeGroupEntityString(toGroupId(bucketName,WorkspaceAccessLevel.Read))
-          val writersEntity = makeGroupEntityString(toGroupId(bucketName,WorkspaceAccessLevel.Write))
-          val ownersEntity = makeGroupEntityString(ownersGroupId)
-          val bucket = new Bucket().setName(bucketName).setAcl(List(
-              newBucketAccessControl(readersEntity,WorkspaceAccessLevel.Read),
-              newBucketAccessControl(writersEntity,WorkspaceAccessLevel.Write),
-              newBucketAccessControl(ownersEntity,WorkspaceAccessLevel.Owner)
-            )).setDefaultObjectAcl(List(
-              newObjectAccessControl(readersEntity,WorkspaceAccessLevel.Read),
-              // NB: writers have read access to objects -- there is no write access, objects are immutable
-              newObjectAccessControl(writersEntity,WorkspaceAccessLevel.Read/*sic!*/),
-              newObjectAccessControl(ownersEntity,WorkspaceAccessLevel.Owner)
-            ))
-          val inserter = getStorage(getBucketCredential(userInfo)).buckets.insert(projectId,bucket)
-          retry(inserter.execute,when500)
+          val ownerInserter = directory.members.insert(ownersGroupId, owner)
+          blocking {
+            ownerInserter.execute
+          }
+        }
+      }
+    }
+
+    def insertBucket: (Member) => Future[Unit] = { _ =>
+      retry(when500) {
+        () => Future {
+          val bucketAcls = groupAccessLevelsAscending.map(ac => newBucketAccessControl(makeGroupEntityString(toGroupId(bucketName, ac)), ac))
+          val defaultObjectAcls = groupAccessLevelsAscending.map(ac => {
+            // NB: writers have read access to objects -- there is no write access, objects are immutable
+            newObjectAccessControl(makeGroupEntityString(toGroupId(bucketName, ac)), if (ac == WorkspaceAccessLevel.Owner) WorkspaceAccessLevel.Owner else WorkspaceAccessLevel.Read)
+          })
+
+          val bucket = new Bucket().setName(bucketName).setAcl(bucketAcls).setDefaultObjectAcl(defaultObjectAcls)
+          val inserter = getStorage(getBucketCredential(userInfo)).buckets.insert(projectId, bucket)
+          blocking {
+            inserter.execute
+          }
         }
       }
     }
+
+    val doItAll = insertGroups(workspaceName, bucketName) flatMap assertSuccessfulGroupInserts flatMap insertOwnerMember flatMap insertBucket
+    doItAll.transform(f => f, rollbackGroups)
```

Comment:
I believe this is still not quite right.  The problem is that insertGroups now <em>always</em> succeeds.